File: src/main/java/me/andre111/d20server/command/TemplateCommand.java
package me.andre111.d20server.command;

import me.andre111.d20common.model.chat.ChatEntry;
import me.andre111.d20common.model.profile.Profile;
import me.andre111.d20common.scripting.ScriptException;
import me.andre111.d20server.service.ChatService;
import me.andre111.d20server.template.Template;

public class TemplateCommand extends Command {
	private final boolean showPublic;
	private final boolean showSelf;
	private final boolean showGM;
	
	public TemplateCommand(String name, String[] aliases, boolean showPublic, boolean showSelf, boolean showGM) {
		super(name, aliases);
		
		this.showPublic = showPublic;
		this.showSelf = showSelf;
		this.showGM = showGM;
	}

	@Override
	public void execute(Profile profile, String arguments) {
		String[] split = arguments.split(" ", 2);
		if(split.length != 2) {
			ChatService.appendError(profile, "Usage: /template <name> <argument>[;<argument>[;...]]");
			return;
		}
		
		String name = split[0].toLowerCase();
		String templateArguments = split[1];
		
		// find template
		Template template = Template.getTemplate(name);
		if(template == null) {
			ChatService.appendError(profile, "Unknown template: "+name);
			return;
		}
		
		// parse template
		String parsed = "";
		try {
			String[] inputs = templateArguments.split(";", -1);
			for(int i=0; i<inputs.length; i++) inputs[i] = ChatService.escape(inputs[i]);
			
			parsed = template.parse(profile, inputs);
		} catch (ScriptException e) {
			ChatService.appendError(profile, "Template parsing error: "+e.getMessage());
			return;
		}
		
		// build message
		StringBuilder sb = new StringBuilder();
		if(ChatService.USE_HTML) {
			sb.append("<p class=\"chat-sender\">");
			sb.append(ChatService.escape(profile.getName()));
			sb.append(": ");
			sb.append("</p>");
			
			sb.append("<p class=\"chat-message\">");
			sb.append(parsed);
			sb.append("</p>");
		} else {
			sb.append(ChatService.STYLE_SENDER);
			sb.append(profile.getName());
			sb.append(": \n");
			sb.append(parsed);
		}
		
		// determine recipents
		long[] recipents = buildRecipents(profile, showPublic, showSelf);
		
		// append message
		ChatService.append(true, new ChatEntry(sb.toString(), profile.id(), showGM, recipents));
	}
}

File: core/server/command/template-command.js
import { ChatEntry } from '../../common/message/chat/chat-entry.js';
import { I18N } from '../../common/util/i18n.js';
import { splitArguments } from '../../common/util/stringutil.js';
import { ChatService } from '../service/chat-service.js';
import { Templates } from '../template/templates.js';
import { Command } from './command.js';

export class TemplateCommand extends Command {
    showPublic;
    showSelf;
    showGM;

    constructor(name, aliases, showPublic, showSelf, showGM) {
        super(name, aliases, false);

        this.showPublic = showPublic;
        this.showSelf = showSelf;
        this.showGM = showGM;
    }

    execute(profile, args) {
        const split = splitArguments(args, 2);
        if (split.length < 2) throw new Error(I18N.get('commands.error.arguments', 'Wrong argument count: %0', '<name> <argument>[;<argument>[;...]]'));

        const name = split[0].toLowerCase();
        const templateArguments = split[1];

        // find template
        const template = Templates.get(name);
        if (!template) throw new Error(I18N.get('command.template.error.unknown', 'Unknown template: %0', name));

        // parse template
        var inputs = templateArguments.split(';');
        for (var i = 0; i < inputs.length; i++) inputs[i] = ChatService.escape(inputs[i]);
        const parseResult = template.parse(profile, inputs);

        // build message
        var text = '<div class="chat-sender">' + ChatService.escape(profile.getUsername()) + ': </div>';
        text = text + '<div class="chat-message">' + parseResult.string + '</div>';

        // determine recipents
        const recipents = this.buildRecipents(profile, this.showPublic, this.showSelf);

        // create entry
        const entry = new ChatEntry(text, profile.getID(), this.showGM, recipents);
        entry.setRolls(parseResult.diceRolls);
        entry.setTriggeredContent(parseResult.triggeredContent);

        // append message
        ChatService.append(true, entry);
    }
}
